contracts/libraries/LibArray.sol

```diff
@@ -0,0 +1,315 @@
+
+      inplaceSortByValue(ids, values);
+      inplaceResize(ids, k);
+
+      return (ids, length - k);
+    }
+  }
+
+  /**
+   * @dev Filter an array of address given array of indices.
+   * @param self The array of address.
+   * @param indices The array of indices to filter.
+   * @return result The resulting array of address.
+   */
+  function take(address[] memory self, uint256[] memory indices) internal pure returns (address[] memory result) {
+    return toAddresses(take(toUint256s(self), indices));
+  }
+
+  /**
+   * @dev Filters the given array of keys based on the provided filter function.
+   * @param keys The array of keys to filter.
+   * @param v The array of values to pass to the filter function.
+   * @param filterFn The filter function that takes a key and an array of values and returns a boolean.
+   * @return filteredIds The array of filtered keys.
+   * @return nFiltered The number of filtered keys.
+   */
+  function filter(
+    uint256[] memory keys,
+    uint256[] memory v,
+    function(uint256, uint256[] memory) pure returns (bool) filterFn
+  ) internal pure returns (uint256[] memory filteredIds, uint256 nFiltered) {
+    unchecked {
+      uint256 length = keys.length;
+      filteredIds = new uint256[](length);
+
+      for (uint256 i; i < length; ++i) {
+        if (filterFn(i, v)) {
+          filteredIds[nFiltered] = keys[i];
+          ++nFiltered;
+        }
+      }
+
+      inplaceResize(filteredIds, nFiltered);
+
+      return (filteredIds, nFiltered);
+    }
+  }
+
+  function inplaceSortByValue(
+    address[] memory self,
+    uint256[] memory values
+  ) internal pure returns (address[] memory sorted) {
+    return toAddresses(inplaceQuickSortByValue(toUint256s(self), values));
+  }
+
+  /**
+   * @dev inplace resizes the given array to the specified length.
+   * @param self The array to be resized.
+   * @param length The new length of the array.
+   * @return resized The resized array.
+   * @notice This function modify `self`
+   */
+  function inplaceResize(uint256[] memory self, uint256 length) internal pure returns (uint256[] memory resized) {
+    assembly ("memory-safe") {
+      resized := self
+      mstore(resized, length)
+    }
+  }
+
+  /**
+   * @dev inplace resizes the given array to the specified length.
+   * @param self The array to be resized.
+   * @param length The new length of the array.
+   * @return resized The resized array.
+   */
+  function inplaceResize(address[] memory self, uint256 length) internal pure returns (address[] memory resized) {
+    return toAddresses(inplaceResize(toUint256s(self), length));
+  }
+
+  /**
+   * @dev Sorts an array of uint256 values based on a corresponding array of values using the specified sorting mode.
+   * @param self The array to be sorted.
+   * @param values The corresponding array of values used for sorting.
+   * @notice This function modify `self` and `values`
+   * @return sorted The sorted array.
```

Comment:
Please improve the docs. Reference: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Arrays.sol#L21-L23